concurrent.futures: allow as_completed to act covariantly

hauntsaninja opened the pull request:
Fixes #11285. I think this works? Not quite sure the best way to describe this, feel free to edit PR title.

Akuli commented on stdlib/concurrent/futures/_base.pyi:
```diff
@@ -69,7 +75,17 @@ class Executor:
+class _AsCompletedFuture(Protocol[_T_co]):
+    # as_completed only mutates non-generic aspects of passed Futures and does not do any nominal
+    # checks. Therefore, we can use a Protocol here to allow as_completed to act covariantly.
+    # See check_concurrent_futures.py
```
File names tend to get outdated when refactoring.
```suggestion
    # There are tests to ensure that this works.
```

Akuli commented on test_cases/stdlib/check_concurrent_futures.py:
```diff
@@ -0,0 +1,23 @@
+    ...
+
+
+def check_as_completed_covariance() -> None:
```
As far as I can tell, the original PR that just makes `Future` covariant would pass this test. The test basically checks that `Future` acts covariantly, and that's all it does.

Can you add a test that the previous PR fails and this PR passes? If you can't, or if the test becomes ends up so "niche" it doesn't do anything practical, I think it's a clear sign that we should just make the thing covariant. My instinct is that it's probably what we should do anyway, because it is practical and simple, but a good test case could convince me otherwise.

Akuli commented:
> feel free to edit PR title.

IMO your title is actually pretty good :)

hauntsaninja commented:
Thanks, I addressed feedback. Let me know what you think of the new test case. I've used `set_result` in real world code before, happy to describe use cases.
